### PR TITLE
Add support for Vue "Generic Components"

### PR DESCRIPTION
**Pull request by auvred**
## Description

Closes https://github.com/prettier/prettier/issues/14864

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Review by codenameryan — APPROVED** [review]

File: src/language-html/embed/vue-attributes.js [hunk 1]
```diff
@@ -27,6 +28,17 @@ function printVueAttribute(path, options) {
     return printVueVForDirective;
   }
 
+  if (
+    attributeName === "generic" &&
+    path.match(
+      undefined,
+      (node) => node.name === "script",
+      (node) => node.type === "root",
```
Comment (fisker) [inline src/language-html/embed/vue-attributes.js:36]: We can use existing `isVueScriptTag` function.

File: src/language-html/embed/vue-script-generic.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { formatAttributeValue } from "./utils.js";
+
+/**
+ * @typedef {import("../../document/builders.js").Doc} Doc
+ */
+
+/**
+ * @returns {Promise<Doc>}
+ */
+function printVueScriptGeneric(textToDoc, print, path) {
```
Comment (fisker) [inline src/language-html/embed/vue-script-generic.js:11]: Maybe `printVueScriptGenericAttributeValue` is more specific.

File: src/language-js/print/estree.js [hunk 3]
```diff
@@ -101,7 +102,11 @@ function printEstree(path, options, print, args) {
     case "JsonRoot":
       return [print("node"), hardline];
     case "File":
-      return printHtmlBinding(path, options, print) ?? print("program");
+      return (
+        printHtmlBinding(path, options, print) ??
+        printVueScriptGenericParameters(path, options, print) ??
```
Comment (fisker) [inline src/language-js/print/estree.js:107]: Let's put it in `printHtmlBinding` like function parameters do.

File: tests/format/vue/vue-3/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -1,5 +1,115 @@
...
+     * comment 2
+     */
+    U extends number, // comment 3
+    /** comment 4 */ C extends MyType
+  "
+></script>
+
+<template>
+  <!-- should not format it here -->
+  <scriptt
```
Comment (fisker) [inline tests/format/vue/vue-3/__snapshots__/jsfmt.spec.js.snap:103]: Let's use `not-script` tag to test.

File: src/language-js/print/type-parameters.js [hunk 5]
```diff
@@ -189,4 +189,29 @@ function printTypeParameter(path, options, print) {
...
+    "body",
+    0,
+    "typeParameters",
+    "params",
+  );
+
+  return group([
+    indent([softline, join([",", line], parameterDocs)]),
+    softline,
+  ]);
```
Comment (fisker) [inline src/language-js/print/type-parameters.js:209]: Indentation should be added in `shouldHugJsExpression`

File: src/language-js/print/estree.js (hunk 3 above)
Comment (auvred) [reply src/language-js/print/estree.js]: I hope I understood it properly, I moved whole `printVueScriptGenericParameters` inside `printHtmlBinding`. Is this right?

Comment (fisker) [reply src/language-js/print/estree.js]: That's correct, thank you.

File: src/language-html/embed/vue-script-generic.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    textToDoc,
+    {
+      parser: "babel-ts",
+      __isVueScriptGeneric: true,
+    },
+    shouldHugJsExpression,
+  );
+}
+
+export { printVueScriptGenericAttributeValue };
```
Comment (fisker) [inline src/language-html/embed/vue-script-generic.js:27]: Can you rename this file to match function name?

File: src/language-html/embed/vue-attributes.js [hunk 7]
```diff
@@ -27,6 +29,13 @@ function printVueAttribute(path, options) {
     return printVueVForDirective;
   }
 
+  if (
+    attributeName === "generic" &&
+    path.match(undefined, (node) => isVueScriptTag(node, options))
```
Comment (fisker) [inline src/language-html/embed/vue-attributes.js:34]:
```suggestion
    isVueScriptTag(node.parent, options)
```

**Review by fisker — APPROVED** [review]
LGTM, thank you!

File: src/language-html/embed/print-vue-script-generic-attribute-value.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  const { node } = path;
+
+  const value = getUnescapedAttributeValue(node);
+
+  return formatAttributeValue(
+    `type T<${value}> = any`,
+    textToDoc,
+    {
+      parser: "babel-ts",
+      __isVueScriptGeneric: true,
```
Comment (fisker) [inline src/language-html/embed/print-vue-script-generic-attribute-value.js:21]: Do you think we should use a better name for this flag? Plugins like `prettier-plugin-pug` may use it in future, we'll not able to change it.

Comment (fisker) [reply src/language-html/embed/print-vue-script-generic-attribute-value.js:21]: Sounds good.

Comment (auvred) [reply src/language-html/embed/print-vue-script-generic-attribute-value.js]: Okay, if we can rename it after, this is fine! Let's leave it as it is. If I need to fix anything else, let me know :)

**Review by fisker — APPROVED** [review]

**Review by sosukesuzuki — APPROVED** [review]
One changelog comment

File: changelog_unreleased/vue/15066.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#### Add support for Vue "Generic Components" (#15066 by @auvred)
+
```
[the remark this answers is not in the document]
Comment (auvred) [reply changelog_unreleased/vue/15066.md:2]: Done :)

**Pull request merged**